packages/merkleize-metadata/src/main.ts

```diff
@@ -77,10 +80,46 @@ export interface MetadataMerkleizer {
 
 export const merkleizeMetadata = (
   metadataBytes: Uint8Array | HexString,
-  info: ExtraInfo,
+  {
+    decimals,
+    tokenSymbol,
+    ...hinted
+  }: { decimals: number; tokenSymbol: string } & Partial<ExtraInfo>,
 ): MetadataMerkleizer => {
   const metadata = getMetadata(metadataBytes)
+  const { ss58Prefix, buildDefinition } = getDynamicBuilder(
+    getLookupFn(metadata),
+  )
+  if (ss58Prefix == null) throw new Error("SS58 prefix not found in metadata")
+  if (hinted.base58Prefix != null && ss58Prefix !== hinted.base58Prefix)
+    throw new Error(
+      `SS58 prefix not expected. Received ${hinted.base58Prefix} expected ${ss58Prefix}`,
+    )
+  const version = metadata.pallets
+    .find((x) => x.name === "System")
+    ?.constants.find((x) => x.name === "Version")
+  if (version == null) throw new Error("System.Version constant not found")
+  const { spec_name: specName, spec_version: specVersion } = buildDefinition(
+    version.type,
+  ).dec(version.value)
+  if (typeof specName !== "string" || typeof specVersion !== "number")
+    throw new Error("Spec name or spec version not found")
+  if (hinted.specName != null && hinted.specName !== specName)
+    throw new Error(
+      `Spec name not expected. Received ${hinted.specName} expected ${specName}`,
+    )
+  if (hinted.specVersion != null && hinted.specVersion !== specVersion)
+    throw new Error(
+      `Spec version not expected. Received ${hinted.specVersion} expected ${specVersion}`,
+    )
```

Comment:
A helper function for this logic would help at decreasing the bundle size. Eg:

```ts
const assertExpected = <T>(name: string, expected: T, received?: T): void => {
  if (received != null && received !== expected)
    throw new Error(`${name} not expected. Received ${received} expected ${expected}`)
}
```

Mostly because of the string errors.

Reply:
Sounds good 👍🏻